pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+			if err != nil {
+				return err
+			}
+			values.deltaEndpointNonce = nonce
+			setState(resp, resource.EndpointType)
+		case resp, more := <-values.deltaClusters:
+			if !more {
+				return status.Errorf(codes.Unavailable, "clusters watch failed")
+			}
+			nonce, err := send(resp, resource.ClusterType)
+			if err != nil {
+				return err
+			}
+			values.deltaClusterNonce = nonce
+			setState(resp, resource.ClusterType)
+		case resp, more := <-values.deltaRoutes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "routes watch failed")
+			}
+			nonce, err := send(resp, resource.RouteType)
+			if err != nil {
+				return err
+			}
+			values.deltaRouteNonce = nonce
+			setState(resp, resource.RouteType)
+		case resp, more := <-values.deltaListeners:
+			if !more {
+				return status.Errorf(codes.Unavailable, "listeners watch failed")
+			}
+			nonce, err := send(resp, resource.ListenerType)
+			if err != nil {
+				return err
+			}
+			values.deltaListenerNonce = nonce
+			setState(resp, resource.ListenerType)
+		case resp, more := <-values.deltaSecrets:
+			if !more {
+				return status.Errorf(codes.Unavailable, "secrets watch failed")
+			}
+			nonce, err := send(resp, resource.SecretType)
+			if err != nil {
+				return err
+			}
+			values.deltaSecretNonce = nonce
+			setState(resp, resource.SecretType)
+		case resp, more := <-values.deltaRuntimes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "runtimes watch failed")
+			}
+			nonce, err := send(resp, resource.RuntimeType)
+			if err != nil {
+				return err
+			}
+			values.deltaRuntimeNonce = nonce
+			setState(resp, resource.RuntimeType)
+		case resp, more := <-values.deltaResponses:
+			if more {
+				if resp == deltaErrorResponse {
+					return status.Errorf(codes.Unavailable, "delta resource watch failed")
+				}
+				typeURL := resp.GetDeltaRequest().TypeUrl
+				nonce, err := send(resp, typeURL)
+				if err != nil {
+					return err
+				}
+				values.deltaNonces[typeURL] = nonce
+				setState(resp, typeURL)
+			}
+		case req, more := <-reqCh:
+			// input stream ended or errored out
+			if !more {
+				return nil
+			}
+			if req == nil {
+				return status.Errorf(codes.Unavailable, "empty request")
+			}
+
+			// Log out our error detail from  envoyif we get one but don't do anything crazy here yet
+			// TODO: embed a logger in the server so we can be a bit more verbose when needed
+			if req.ErrorDetail != nil {
+				return status.Errorf(codes.Code(req.ErrorDetail.GetCode()), "received error from xDS client: %s", req.ErrorDetail.GetMessage())
+			}
+
+			// node field in discovery request is delta-compressed
+			// nonces can be reused across streams; we verify nonce only if nonce is not initialized
+			var nonce string
+			if req.Node != nil {
+				node = req.Node
+				nonce = req.GetResponseNonce()
+			} else {
+				req.Node = node
+				// If we have no nonce, i.e. this is the first request on a delta stream, set one
```

Comment:
I don't see a conditional here? Seems like we always set this

Reply:
I guess that's true, we set that only while looking at the `req.Node`. Does envoy continually send over the node information after the initial xDS request? I think maybe that's what they were going off of in SOTW when they wrote this? That would be my only guess

Reply:
I think the comment here is incorrect but the code is doing the right thing. Envoy can send node info only on the first request in the stream if we set `set_node_on_first_message_only: true` So in this case we just assume we are using the last nonce that was used in the current stream.

Reply:
I think that's correct. Which means the reused nonce is picked up on on a reconnect or something of that nature

Reply:
No, definitely not. Nonces are only valid in the context of the stream in which they are sent. Versions persist, e.g. in `initial_resource_versions`, but not nonce.

Reply:
Oops sorry I meant "aren't nonces not supposed to live past streams"! You're definitely right there